feat(scan): define the scan state and issues list retrieval API

ostridm opened the pull request:
closes #49

ostridm committed:
feat(scan): define the scan state and issues list retrieval API

closes #49

derevnjuk commented on src/SecTester.Scan/Scan.cs:
```diff
@@ -1,58 +1,115 @@
   public Task Stop()
   {
-    throw new NotImplementedException();
+    return _scans.StopScan(Id);
```
You have to make sure that the scan is not stopped. To receive an actual state, you have to call `RefreshState` manually beforehand. Since the state is shared between threads, you have to use a lock, semaphore for example

ostridm replied:
@derevnjuk What do you think if the `Stop` itself may be called concurrently? Thus having the latest state from `App` is worthless because of race conditions and may also lead to an unwanted call of `Scans.StopScan`.

derevnjuk replied:
The comment above gives you a precise answer to this question. To get into the race conditions using a lock, someone should refresh the state to done-like status when another thread is between `Active` and `Stop` expressions. That is why you have to silently handle such exceptions. But it is not possible to get into this bug just by calling `Stop` concurrently on the condition that one of the threads acquires a lock.

derevnjuk commented on src/SecTester.Scan/Scan.cs:
```diff
@@ -1,58 +1,115 @@
+    {
+      await RefreshState().ConfigureAwait(false);
+
+      if (!Active)
```
ditto

derevnjuk commented on src/SecTester.Scan/Scan.cs:
```diff
@@ -1,58 +1,115 @@
+  private volatile ScanState _scanState = new(ScanStatus.Pending);
 
-  public Scan(Scans scans, ScanOptions options)
+  public string Id { get; init; }
```
Make property get-only

```suggestion
  public string Id { get; }
```

derevnjuk commented on src/SecTester.Scan/Scan.cs:
```diff
@@ -1,58 +1,115 @@
+  private readonly ILogger _logger;
   private readonly Scans _scans;
-  private readonly ScanState _scanState = new(ScanStatus.Pending);
+  private volatile ScanState _scanState = new(ScanStatus.Pending);
```
Using volatile you can easily get wrong. You can break the whole flow because there is no guarantee that the state is the latest possible and methods such as `Delete` and `Stop` are error-prone

derevnjuk commented on test/SecTester.Scan.Tests/ScanTests.cs:
```diff
@@ -1,53 +1,228 @@
 namespace SecTester.Scan.Tests;
 
-public class ScanTests
+public class ScanTests : ScanFixture
```
Instead of inheritance, you have to use composition (see `IClassFixture`). `XUnit` is smart enough to handle their life circle by themself

ostridm replied:
@derevnjuk I do not see any need to use `IClassFixture` since these tests do not require sharing any expensive context and moreover require isolation from each other. I will rework the code to get rid of inheritance

derevnjuk replied:
@ostridm so, then it is not clear why you create a shared `ScanFixture` if you need an isolated context. This is a broken design.

derevnjuk commented on test/SecTester.Scan.Tests/ScanTests.cs:
```diff
@@ -1,53 +1,228 @@
+using Microsoft.Extensions.Logging;
+using NSubstitute.ClearExtensions;
```
Utilize global using, I have already pointed out this twice

derevnjuk commented on test/SecTester.Scan.Tests/Fixtures/ScanFixture.cs:
```diff
@@ -1,20 +1,19 @@
 using SecTester.Core.Bus;
-using SecTester.Scan.CI;
-using SecTester.Scan.Models;
+using Request = SecTester.Scan.Models.Request;
```
ditto

derevnjuk commented on test/SecTester.Scan.Tests/ScanTests.cs:
```diff
@@ -1,53 +1,228 @@
+  public void Constructor_GivenNullId_ThrowError()
+  {
+    // act
+    Action act = () => new Scan(null!, _scans, _logger, new ScanOptions());
```
Use var instead (the same in other places)

```suggestion
    var act = () => new Scan(null!, _scans, _logger, new ScanOptions());
```

derevnjuk replied:
As discussed, we don't waste time testing logic-less constructors. You can keep these tests, but next time avoid writing them.

derevnjuk commented on test/SecTester.Scan.Tests/ScanTests.cs:
```diff
@@ -1,53 +1,228 @@
+    var runningScanState = new ScanState(ScanStatus.Running);
+    var stoppedScanState = new ScanState(ScanStatus.Stopped);
+
+    await using var scan = new Scan(ScanId, _scans, _logger, new ScanOptions(PollingInterval: TimeSpan.Zero));
```
Create a scan instance once in the constructor

derevnjuk commented on test/SecTester.Scan.Tests/ScanTests.cs:
```diff
@@ -1,53 +1,228 @@
+  }
+
+  [Fact]
+  public async Task Status_ReturnsState()
```
```suggestion
  public async Task Status_ReturnsStatus()
```

derevnjuk replied:
I don't see any test to verify that `Status` return the last obtained status when the status is done

derevnjuk commented on test/SecTester.Scan.Tests/ScanTests.cs:
```diff
@@ -1,53 +1,228 @@
 
   [Fact]
-  public void Active_DefaultValue_ReturnTrue()
+  public async Task Status_GivenCancelledToken_ThrowError()
```
```suggestion
  public async Task Status_PollingCanceled_ReThrowsError()
```

derevnjuk commented on test/SecTester.Scan.Tests/ScanTests.cs:
```diff
@@ -1,53 +1,228 @@
     // arrange
-    var scans = Substitute.For<Scans>();
+    var runningScanState = new ScanState(ScanStatus.Running);
+    var stoppedScanState = new ScanState(ScanStatus.Stopped);
```
The last status must be running to verify if it actually works

derevnjuk commented on test/SecTester.Scan.Tests/ScanTests.cs:
```diff
@@ -1,53 +1,228 @@
 
   [Fact]
-  public void Done_DefaultValue_ReturnTrue()
+  public async Task Status_SwitchingDoneState_StopsYield()
```
The test case is not clear
```suggestion
  public async Task Status_FinalStateReached_StopsQueryingState()
```

derevnjuk commented on test/SecTester.Scan.Tests/ScanTests.cs:
```diff
@@ -1,53 +1,228 @@
+  }
+
+  [Fact]
+  public async Task Status_GivenQueuedStateConsequently_LogsConcurrencyWarning()
```
```suggestion
  public async Task Status_EntersIntoQueued_LogsConcurrencyWarning()
```

derevnjuk commented on test/SecTester.Scan.Tests/ScanTests.cs:
```diff
@@ -1,53 +1,228 @@
+  }
+
+  [Fact]
+  public async Task Status_GivenAnyStateAfterQueued_LogsResumingExecution()
```
```suggestion
  public async Task Status_Dequeued_LogsResumingExecution()
```

derevnjuk commented on test/SecTester.Scan.Tests/ScanTests.cs:
```diff
@@ -1,53 +1,228 @@
+    await using var scan = new Scan(ScanId, _scans, _logger, new ScanOptions(PollingInterval: TimeSpan.Zero));
+
+    _scans.GetScan(ScanId).Returns(Task.FromResult(scanState),
+      Task.FromResult(scanState));
```
BTW you can omit `Task.FromResult`

```suggestion
      scanState);
```

derevnjuk commented on test/SecTester.Scan.Tests/ScanTests.cs:
```diff
@@ -1,53 +1,228 @@
+  }
+
+  [Fact]
+  public async Task Status_GivenAnyStateAfterQueued_LogsResumingExecution()
```
Please cover the case when `Status` stops consuming while receiving the done-like statuses

ostridm replied:
`Status_SwitchingDoneState_StopsYield` is responsible for this
https://github.com/NeuraLegion/sectester-net/pull/59#discussion_r1035229969

derevnjuk replied:
This test covers only a single case with `done` when you have two more states that lead to the same behavior. BTW never a close conversation which requires my attention.

derevnjuk commented on test/SecTester.Scan.Tests/ScanTests.cs:
```diff
@@ -1,53 +1,228 @@
+  }
+
+  [Fact]
+  public async Task DisposeAsync_ActiveIsFalse_DeletesScan()
```
```suggestion
  public async Task DisposeAsync_ScanIsStopped_DeletesFromApp()
```

derevnjuk replied:
Add a case to cover an error while deleting a scan from the app

derevnjuk commented on test/SecTester.Scan.Tests/ScanTests.cs:
```diff
@@ -1,53 +1,228 @@
+  }
+
+  [Fact]
+  public async Task DisposeAsync_ActiveIsTrue_KeepsScan()
```
```suggestion
  public async Task DisposeAsync_ScanIsActive_DoesNothing()
```

derevnjuk commented on test/SecTester.Scan.Tests/SecTester.Scan.Tests.csproj:
```diff
@@ -10,6 +10,7 @@
   </ItemGroup>
 
   <ItemGroup>
+    <PackageReference Include="System.Linq.Async" Version="6.0.1" />
```
Do you use it somewhere?

ostridm replied:
Use to test aync enumerable `<async enumerable>.LastOrDefaultAsync();`

ostridm committed:
feat(scan): define the scan state and issues list retrieval API

adjust according to code review
https://github.com/NeuraLegion/sectester-net/pull/59#discussion_r1035188908
https://github.com/NeuraLegion/sectester-net/pull/59#discussion_r1035191702
https://github.com/NeuraLegion/sectester-net/pull/59#discussion_r1035194488
https://github.com/NeuraLegion/sectester-net/pull/59#discussion_r1035201131
https://github.com/NeuraLegion/sectester-net/pull/59#discussion_r1035208599
https://github.com/NeuraLegion/sectester-net/pull/59#discussion_r1035209293
https://github.com/NeuraLegion/sectester-net/pull/59#discussion_r1035209375
https://github.com/NeuraLegion/sectester-net/pull/59#discussion_r1035214415
https://github.com/NeuraLegion/sectester-net/pull/59#discussion_r1035215806
https://github.com/NeuraLegion/sectester-net/pull/59#discussion_r1035219597
https://github.com/NeuraLegion/sectester-net/pull/59#discussion_r1035222803
https://github.com/NeuraLegion/sectester-net/pull/59#discussion_r1035223447
https://github.com/NeuraLegion/sectester-net/pull/59#discussion_r1035228610
https://github.com/NeuraLegion/sectester-net/pull/59#discussion_r1035229969
https://github.com/NeuraLegion/sectester-net/pull/59#discussion_r1035233549
https://github.com/NeuraLegion/sectester-net/pull/59#discussion_r1035234824
https://github.com/NeuraLegion/sectester-net/pull/59#discussion_r1035236178
https://github.com/NeuraLegion/sectester-net/pull/59#discussion_r1035237203
https://github.com/NeuraLegion/sectester-net/pull/59#discussion_r1035238084
https://github.com/NeuraLegion/sectester-net/pull/59#discussion_r1035238330
https://github.com/NeuraLegion/sectester-net/pull/59#discussion_r1035239486
https://github.com/NeuraLegion/sectester-net/pull/59#discussion_r1035242747

closes #49

derevnjuk commented on src/SecTester.Scan/Internal/SemaphoreLock.cs:
```diff
@@ -0,0 +1,39 @@
+
+namespace SecTester.Scan.Internal;
+
+internal class SemaphoreLock : IDisposable
```
Your implementation is not safe, see #75

ostridm replied:
Copied yours. But why don't put it to `Core`?

derevnjuk replied:
This is a question to you, no to me. It was designed to be used internally in the repeater project, if you need to do the same, feel free to extract it into the core project.

derevnjuk commented on src/SecTester.Scan/Scan.cs:
```diff
@@ -1,58 +1,164 @@
-    throw new NotImplementedException();
+    ScanState newState;
+
+    if (_scanStateSemaphore.CurrentCount == 0)
+    {
+      using var _ = await SemaphoreLock.LockAsync(_scanStateSemaphore, cancellationToken).ConfigureAwait(false);
+      newState = _scanState;
+    }
```
Well, if `CurrentCount` is 0, this means the lock has already been acquired

ostridm replied:
Changed to flag. A lock may be acquired by `Active` and `Done` getters.

ostridm committed:
feat(scan): define the scan state and issues list retrieval API

adjust according to code review
https://github.com/NeuraLegion/sectester-net/pull/59#discussion_r1036995554
https://github.com/NeuraLegion/sectester-net/pull/59#discussion_r1037000957

closes #49

derevnjuk commented on src/SecTester.Scan/Extensions/SemaphoreSlimExtensions.cs:
```diff
@@ -0,0 +1,37 @@
+
+namespace SecTester.Scan.Extensions;
+
+internal static class SemaphoreSlimExtensions
```
https://github.com/NeuraLegion/sectester-net/pull/59#discussion_r1037471312

derevnjuk commented on src/SecTester.Scan/Internal/AutoReleasableSmaphore.cs:
```diff
@@ -0,0 +1,28 @@
+
+namespace SecTester.Scan.Internal;
+
+internal class AutoReleasableSemaphore : IDisposable
```
ditto

derevnjuk commented on src/SecTester.Scan/Scan.cs:
```diff
@@ -1,58 +1,165 @@
-  private readonly ScanState _scanState = new(ScanStatus.Pending);
+  private int _gettingScanState;
+  private ScanState _scanState = new(ScanStatus.Pending);
+  private readonly SemaphoreSlim _scanStateSemaphore = new(1, 1);
```
```suggestion
  private readonly SemaphoreSlim _semaphore = new(1, 1);
```

derevnjuk commented on src/SecTester.Scan/Scan.cs:
```diff
@@ -1,58 +1,165 @@
+    {
+      using var _ = await _scanStateSemaphore.LockAsync(cancellationToken).ConfigureAwait(false);
+      return _scanState;
+    }
```
It looks like a weird optimization. Why don't give a thread to refresh the state once the lock is released?

derevnjuk commented on src/SecTester.Scan/Scan.cs:
```diff
@@ -1,58 +1,165 @@
+      return _scanState;
+    }
+
+    using (await _scanStateSemaphore.LockAsync(cancellationToken).ConfigureAwait(false))
```
Remove brackets from `using` statement.

derevnjuk commented on src/SecTester.Scan/Scan.cs:
```diff
@@ -1,58 +1,165 @@
+
+    using (await _scanStateSemaphore.LockAsync(cancellationToken).ConfigureAwait(false))
+    {
+      Interlocked.Increment(ref _gettingScanState);
```
Using interlocked or any kind of memory barrier is pointless once you have a lock

derevnjuk commented on src/SecTester.Scan/Scan.cs:
```diff
@@ -1,58 +1,165 @@
+      await Task.Delay(_options.PollingInterval ?? DefaultPollingInterval, cancellationToken).ConfigureAwait(false);
+
+      yield return await RefreshState(cancellationToken).ConfigureAwait(false);
+    }
```
You have to yield return the `ScanState` at the end if you are going to receive a final state once the scan is done

derevnjuk commented on src/SecTester.Scan/Scan.cs:
```diff
@@ -1,58 +1,165 @@
+
+      if (!Active)
+      {
+        await _scans.DeleteScan(Id).ConfigureAwait(false);
```
Please don't remove the scan here) Otherwise, the user won't be able to see results) To implement this behavior, it is better to implement the method `DisposeAsyncCore` https://learn.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-disposeasync#the-disposeasynccore-method with some parameter, e.g. `bool deleteScan`

derevnjuk commented on test/SecTester.Scan.Tests/ScanTests.cs:
```diff
@@ -1,53 +1,324 @@
+  }
+
+  [Fact]
+  public async Task Stop_StopScanThrows_Returns()
```
```suggestion
  public async Task Stop_StopScanThrows_DoesNothing()
```

derevnjuk commented on test/SecTester.Scan.Tests/ScanTests.cs:
```diff
@@ -1,53 +1,324 @@
+  }
+
+  [Fact]
+  public async Task Stop_GetScanThrows_Returns()
```
ditto

derevnjuk commented on test/SecTester.Scan.Tests/ScanTests.cs:
```diff
@@ -1,53 +1,324 @@
 
   [Fact]
-  public void Done_DefaultValue_ReturnTrue()
+  public async Task DisposeAsync_DeleteScanThrows_Returns()
```
ditto

derevnjuk commented on test/SecTester.Scan.Tests/ScanTests.cs:
```diff
@@ -1,53 +1,324 @@
 
   [Fact]
-  public void Active_DefaultValue_ReturnTrue()
+  public async Task DisposeAsync_GetScanThrows_Returns()
```
ditto

derevnjuk commented on test/SecTester.Scan.Tests/Usings.cs:
```diff
@@ -3,14 +3,21 @@
 global using SecTester.Scan.Extensions;
 global using SecTester.Scan.Models;
 global using SecTester.Scan.Target.Har;
+global using SecTester.Scan.Tests.Extensions;
+global using SecTester.Scan.Tests.Fixtures;
+global using SecTester.Scan.Tests.Mocks;
```
Keep them in tests, only external dependencies should be placed in global using

derevnjuk commented on test/SecTester.Scan.Tests/ScanTests.cs:
```diff
@@ -1,53 +1,324 @@
+    _scans.GetScan(ScanFixture.ScanId)
+      .Returns(new ScanState(ScanStatus.Running), scanState, new ScanState(ScanStatus.Running));
+
+    // ADHOC: LastOrDefaultAsync() may hang-up when 'FinalState' condition is not satisfied   
```
See comment above

derevnjuk commented on src/SecTester.Scan/Scan.cs:
```diff
@@ -1,58 +1,165 @@
+  {
+    get
+    {
+      using var _ = _scanStateSemaphore.Lock();
```
It is better to use `ReaderWriterLock` for such shared resources, but actually, I  don't see any reason to do this for non-static resources

ostridm committed:
feat(scan): define the scan state and issues list retrieval API

adjust according to code review
https://github.com/NeuraLegion/sectester-net/pull/59#discussion_r1037916554
https://github.com/NeuraLegion/sectester-net/pull/59#discussion_r1037916651
https://github.com/NeuraLegion/sectester-net/pull/59#discussion_r1038021693
https://github.com/NeuraLegion/sectester-net/pull/59#discussion_r1038025294
https://github.com/NeuraLegion/sectester-net/pull/59#discussion_r1038025774
https://github.com/NeuraLegion/sectester-net/pull/59#discussion_r1038026396
https://github.com/NeuraLegion/sectester-net/pull/59#discussion_r1038122240
https://github.com/NeuraLegion/sectester-net/pull/59#discussion_r1038122286
https://github.com/NeuraLegion/sectester-net/pull/59#discussion_r1038122668
https://github.com/NeuraLegion/sectester-net/pull/59#discussion_r1038122949
https://github.com/NeuraLegion/sectester-net/pull/59#discussion_r1038123595
https://github.com/NeuraLegion/sectester-net/pull/59#discussion_r1038132361
https://github.com/NeuraLegion/sectester-net/pull/59#discussion_r1038029035

closes #49

ostridm committed:
feat(scan): define the scan state and issues list retrieval API

closes #49

derevnjuk commented on src/SecTester.Scan/Scan.cs:
```diff
@@ -1,58 +1,160 @@
   {
+    get
+    {
+      using var _ = _semaphore.Lock();
```
As discussed yesterday, it is pointless to lock this for reading. You just need to hold a lock on the assignment due to the immutable nature of `ScanState`. You just need to read a copy to ensure atomicity.

derevnjuk commented on src/SecTester.Scan/Scan.cs:
```diff
@@ -1,58 +1,160 @@
+  {
+    get
+    {
+      using var _ = _semaphore.Lock();
```
ditto

derevnjuk commented on src/SecTester.Scan/Scan.cs:
```diff
@@ -1,58 +1,160 @@
-    throw new NotImplementedException();
+    try
+    {
+      using var _ = await _semaphore.LockAsync().ConfigureAwait(false);
+
+      await RefreshStateCore().ConfigureAwait(false);
```
You can easily replace this with `RefreshState`. It looks like `RefreshStateCore` is useless

derevnjuk commented on src/SecTester.Scan/ScanOptions.cs:
```diff
@@ -2,8 +2,5 @@
-  public TimeSpan? Timeout { get; set; } = Timeout;
-  public TimeSpan? PollingInterval { get; set; } = PollingInterval;
-}
+public record ScanOptions(TimeSpan? Timeout = default, TimeSpan? PollingInterval = default,
```
Assign meaningful default values to fields, then you can remove `?` operator

derevnjuk commented on test/SecTester.Scan.Tests/Target/TargetTests.cs:
```diff
@@ -1,3 +1,5 @@
+using SecTester.Scan.Target.Har;
```
global using

derevnjuk commented on test/SecTester.Scan.Tests/ScanTests.cs:
```diff
@@ -1,53 +1,350 @@
+using SecTester.Scan.Models;
+using SecTester.Scan.Tests.Mocks;
+using Request = SecTester.Scan.Models.Request;
```
ditto

derevnjuk commented on test/SecTester.Scan.Tests/Models/UploadHarOptionsTests.cs:
```diff
@@ -1,3 +1,6 @@
+using SecTester.Scan.Models;
+using SecTester.Scan.Target.Har;
```
ditto

derevnjuk commented on test/SecTester.Scan.Tests/Models/IssueTests.cs:
```diff
@@ -1,3 +1,4 @@
+using SecTester.Scan.Models;
 using Request = SecTester.Scan.Models.Request;
```
ditto

derevnjuk commented on test/SecTester.Scan.Tests/Extensions/ServiceCollectionExtensionsTests.cs:
```diff
@@ -1,3 +1,8 @@
+using SecTester.Bus.Extensions;
+using SecTester.Core;
+using SecTester.Scan.CI;
+using SecTester.Scan.Extensions;
```
ditto

derevnjuk commented on test/SecTester.Scan.Tests/DefaultScansTests.cs:
```diff
@@ -1,44 +1,104 @@
-using SecTester.Scan.Tests.Fixtures;
+using SecTester.Core;
+using SecTester.Core.Bus;
+using SecTester.Core.Exceptions;
+using SecTester.Scan.CI;
+using SecTester.Scan.Commands;
+using SecTester.Scan.Models;
+using SecTester.Scan.Target.Har;
```
ditto

derevnjuk commented on test/SecTester.Scan.Tests/Commands/UploadHarTests.cs:
```diff
@@ -1,21 +1,29 @@
-using SecTester.Scan.Tests.Fixtures;
+using SecTester.Bus.Dispatchers;
+using SecTester.Scan.Commands;
+using SecTester.Scan.Models;
+using SecTester.Scan.Target.Har;
```
ditto

derevnjuk commented on test/SecTester.Scan.Tests/Commands/CreateScanTests.cs:
```diff
@@ -1,29 +1,54 @@
-using SecTester.Scan.Tests.Fixtures;
+using SecTester.Bus.Dispatchers;
+using SecTester.Scan.Commands;
+using SecTester.Scan.Models;
```
ditto

derevnjuk commented on test/SecTester.Scan.Tests/CI/DefaultCiDiscoveryTests.cs:
```diff
@@ -1,3 +1,5 @@
+using SecTester.Scan.CI;
```
ditto

ostridm committed:
feat(scan): define the scan state and issues list retrieval API

adjust according to code review
https://github.com/NeuraLegion/sectester-net/pull/59#discussion_r1040575508
https://github.com/NeuraLegion/sectester-net/pull/59#discussion_r1040575624
https://github.com/NeuraLegion/sectester-net/pull/59#discussion_r1040577510
https://github.com/NeuraLegion/sectester-net/pull/59#discussion_r1040578960
https://github.com/NeuraLegion/sectester-net/pull/59#discussion_r1040580583
https://github.com/NeuraLegion/sectester-net/pull/59#discussion_r1040580782
https://github.com/NeuraLegion/sectester-net/pull/59#discussion_r1040580881
https://github.com/NeuraLegion/sectester-net/pull/59#discussion_r1040580881
https://github.com/NeuraLegion/sectester-net/pull/59#discussion_r1040580923
https://github.com/NeuraLegion/sectester-net/pull/59#discussion_r1040581044
https://github.com/NeuraLegion/sectester-net/pull/59#discussion_r1040581192
https://github.com/NeuraLegion/sectester-net/pull/59#discussion_r1040581283
https://github.com/NeuraLegion/sectester-net/pull/59#discussion_r1040581376
https://github.com/NeuraLegion/sectester-net/pull/59#discussion_r1040581432

closes #49
